go-controller/cmd/ovnkube/ovnkube.go

```diff
@@ -283,7 +283,13 @@ func runOvnKube(ctx *cli.Context) error {
 	// now that ovnkube master/node are running, lets expose the metrics HTTP endpoint if configured
 	// start the prometheus server to serve OVN K8s Metrics (default master port: 9409, node port: 9410)
 	if config.Kubernetes.MetricsBindAddress != "" {
-		metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof)
+		if config.Kubernetes.OVNMetricsBindAddress == "" {
+			metrics.RegisterOvnMetrics(ovnClientset.KubeClient, master)
+			metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof, true)
+		} else {
+			metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof, false)
+		}
+
 	}
 
```

Comment:
L296-299 also exposes the OVN Metrics on a different Mux in addition to the Node metrics server.

In other words, with this change, the OVN metrics are accessible at:

http://nodeIP:9410/ovnmetrics
http://nodeIP:9476/metrics

Note: We reserved following ports here https://github.com/prometheus/prometheus/wiki/Default-port-allocations

OVN K8s Master/Manager -> 9409
OVN K8s Node/Agent -> 9410
OVN -> 9476
OVS -> 9310

@roytman @palanik1 @squeed What it the reasoning behind this change? Should this target just OpenShift OVN K8s downstream? Thanks.

Reply:
@girishmg Right now OVNMetrics are exposed only if OVNMetricsBindAddress is set in the config. This PR intends to expose OVNMetrics by default (for observability and debugging). Depending on whether OVNMetricsBindAddress is defined, we expose the metrics on either nodeIP:9410/ovnmetrics or nodeIP:9476/metrics.

Reply:
@girishmg Basically, we're just trying to cut down on the number of ports, since we wrap everything with an authorizing proxy.

I mean, we could keep this downstream, but we are trying not to be divergent. What's the reasoning behind using separate ports? Just for differentiatability? Are there duplicate metrics otherwise?

Reply:
@squeed Yes, for differentiation. The thinking was that OVN K8s is layered with OVN K8s, OVN, and OVS components. We wanted to export metrics at each layer at their own port. Initially, they were exported by separate processes. But eventually OVN K8s and OVN got merged into ovnkube worker node implementation.

Will the memory footprint of the ovnkube-node now increase because it needs to now keep the same metrics twice -- one behind the nodeIP:9410/ovnmetrics and an another behind the nodeIP:9476/metrics endpoint. Should we export the metrics one way or the other using an option? That is OVN metrics muxed behind 9410 or 9476 but not both.

Reply:
@girishmg We expose the metrics on a single port. The `config.Kubernetes.OVNMetricsBindAddress` decides whether we expose OVNMetrics behind 9410 *or* 9476.

Reply:
@palanik1 this is the final code:

```
	if config.Kubernetes.MetricsBindAddress != "" {
		if config.Kubernetes.OVNMetricsBindAddress == "" {
			metrics.RegisterOvnMetrics(ovnClientset.KubeClient, master)
			metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof, true)
		} else {
			metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof, false)
		}

	}

	// start the prometheus server to serve OVN Metrics (default port: 9476)
	if config.Kubernetes.OVNMetricsBindAddress != "" {
		metrics.RegisterOvnMetrics(ovnClientset.KubeClient, node)
		metrics.StartOVNMetricsServer(config.Kubernetes.OVNMetricsBindAddress)
	}
```
So, aren't we registering the OVN metrics twice? Behind two different Mux and therefore duplicating the metric collection ?

Reply:
@girishmg Our StartMetricsServer method takes an additional third parameter (enableOvnMetrics) that decides whether we expose ovnmetrics on the MetricsServer Mux or a separate mux.  Let me know if this makes sense.

Reply:
> @palanik1 this is the final code:
> 
> ```
> 	if config.Kubernetes.MetricsBindAddress != "" {
> 		if config.Kubernetes.OVNMetricsBindAddress == "" {
> 			metrics.RegisterOvnMetrics(ovnClientset.KubeClient, master)
> 			metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof, true)
> 		} else {
> 			metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof, false)
> 		}
> 
> 	}
> 
> 	// start the prometheus server to serve OVN Metrics (default port: 9476)
> 	if config.Kubernetes.OVNMetricsBindAddress != "" {
> 		metrics.RegisterOvnMetrics(ovnClientset.KubeClient, node)
> 		metrics.StartOVNMetricsServer(config.Kubernetes.OVNMetricsBindAddress)
> 	}
> ```
> 
> So, aren't we registering the OVN metrics twice? Behind two different Mux and therefore duplicating the metric collection ?

@girishmg , we have 			
`metrics.RegisterOvnMetrics(ovnClientset.KubeClient, master)` wrapped with `config.Kubernetes.OVNMetricsBindAddress == ""` first and then wrapped with  config.Kubernetes.OVNMetricsBindAddress != "" . 
Both can't be true. Does that address the concern? Or did I misunderstand?

Reply:
@girishmg We had introduced a third boolean parameter `enableOvnMetrics` in our StartMetricsServer() method that exposes ovnmetrics on MetricsServer mux only when true.

Reply:
@sdsen thanks for explanation. Sorry, I had overlooked `==` and `!=`.